feat: push notifications new controller

matteoscurati opened the pull request:
## **Description**

This PR introduces a new SW (Service Worker) and a new controller for managing push notifications.

This integration, due to the differences between MV2 and MV3 regarding the use of background scripts and the SW, is only related to the manifest V2. For the manifest V3, part of this logic will be modified, using the extension's own service worker as a starting point. With manifest V2, on the contrary, it is necessary to initialize a dedicated service worker so that FCM (Firebase Cloud Messaging) can handle messages received in the background.

The final UI will allow the user to enable/disable the receipt of notifications. However, this PR concerns only the

new controller and the build management for the creation of the SW.

The new controller will work in accordance with two other controllers for managing authentication and for managing the triggers necessary to receive notifications.

## **Related issues**

No related issues.

Fixes:

## **Manual testing steps**

At the moment, the necessary environment variables have not been inserted into the build process. To test the branch locally, feel free to ask me for the env to insert into your local .metamaskrc file

## **Screenshots/Recordings**

In this video, I show the interaction between a test server for sending notifications and the extension:

https://www.loom.com/share/702fc6a8aea74510acdb1bfac614fa4a

### **Before**

### **After**

Check the entire flow here https://www.figma.com/file/c7GgNw2kScGrVyRGAPhwEd/Notifications?type=design&node-id=484%3A44713&mode=design&t=QsKrmdqZnx42IrMt-1

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

matteoscurati committed:
feat: :sparkles: add fcm feature to handle push notifications in the background

matteoscurati committed:
feat: :sparkles: add a delete token method

matteoscurati committed:
feat: :sparkles: a new controller to hande push platform notifications

matteoscurati committed:
test: :rotating_light: fix verify-locales test

matteoscurati committed:
feat: :green_heart: add push service endpoint

matteoscurati committed:
refactor: :recycle: use async/await to be consistent with the rest of the codebase

matteoscurati committed:
refactor: :pencil2: use a better name to identify the bearer token

danjm commented on app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-app.js:
```diff
@@ -0,0 +1,2735 @@
+/* eslint-disable */
```
@matteoscurati could you include a link to where you copied the code from in this code comment?

matteoscurati replied:
You're right. I'll include the license and the source of these libraries 👍

rekmarks replied:
Can we not use the npm package for this?

matteoscurati replied:
Great question. Unfortunately, Lavamoat node doesn't support ESM, and Firebase does not expose an entry point in CJS. For this reason, at least for now, we had to import Firebase in the way indicated. If I understand correctly, this problem will be solved with the release of Endomoat

FrederikBolding replied:
Can we transpile Firebase using Babel instead? 

There is an example of how we did this in the past here: https://github.com/MetaMask/metamask-extension/pull/17251/files#diff-a05d1f077953d24c0d8ddecf64163da2653ba3b728e1ba1f18af6d770767b127R831-R841

matteoscurati replied:
Ok, thanks to @FrederikBolding 's suggestion, I'm able to import Firebase using npm instead of importing the libraries directly!

danjm commented on app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-messaging-sw.js:
```diff
@@ -0,0 +1,4 @@
+/* eslint-disable */
```
A few questions:
- why does this file need to be minified?
- is it copied from somewhere? if so where? and then does a license need to be included?
- can we include a comment at the top of this file explaining what it is?

danjm replied:
Same for the below file

matteoscurati replied:
Ok, see the comment above.


The library is minified because it is published that way here https://www.gstatic.com/firebasejs/10.8.0/firebase-messaging-sw.js.

In any case, I have proceeded to expand it.

matteoscurati replied:
Sure. See the comment  above.

The library is minified because it is published that way here [https://www.gstatic.com/firebasejs/10.8.0/firebase-messaging-sw.js](https://www.gstatic.com/firebasejs/10.8.0/firebase-messaging-sw.js).

In any case, I have proceeded to expand it.

danjm commented on app/scripts/controllers/push-platform-notifications/firebase/utils/get-notification-data.ts:
```diff
@@ -0,0 +1,94 @@
+  decimalPlaces: 4,
+};
+
+export function calcTokenAmount(value: string, decimals: string) {
```
I see you added a code comment for the same function in shared/lib/transactions-controller-utils.js. Did you intend to import that function here?

danjm commented on shared/constants/platform-notifications.ts:
```diff
@@ -0,0 +1,149 @@
+  DATA_ERC1155 = 'data_erc1155',
+}
+
+export const CHAIN_SYMBOLS = {
```
This const seems to serve a similar purpose as the `CURRENCY_SYMBOLS` or `CHAINLIST_CURRENCY_SYMBOLS_MAP` const in `shared/constants/network.ts`. However, some of the symbols here are different from the symbols in that file (for example, the optimism symbol is different).

I might make sense to import the symbols from `shared/constants/network.ts` to this file, to make sure we are using the same symbols throughout the application.

(Which are the correct symbols is a separate question, but we should make sure we are consistent across the application)

matteoscurati replied:
Yes, you're absolutely right. I've updated the file!

Prithpal-Sooriya replied:
I have remove this file :)

danjm commented on shared/constants/platform-notifications.ts:
```diff
@@ -0,0 +1,149 @@
+  ERC1155_RECEIVED = 'erc1155_received',
+}
+
+const chains = {
```
Nit: because we refer to `ETHEREUM` as `MAINNET` throughout the `shared/constants/network` file, and that is used throughout the codebase, it might be best to name this `chains.ETHEREUM` property as `chains.MAINNET`

danjm commented on app/scripts/controllers/push-platform-notifications/firebase/utils/get-notification-message.ts:
```diff
@@ -0,0 +1,285 @@
+
+const notificationMessageDict: NotificationMessageDict = {
+  erc20_sent: {
+    title: t('pushPlatformNotificationsFundsSentTitle') ?? 'Funds sent',
```
These fallbacks to english strings (e.g. `?? 'Funds sent'`) should not be needed, because the `t` function from the `translate.ts` file falls back to the english translation if no translation in the current locale exists

danjm replied:
the same applied to all cases in this file

matteoscurati replied:
Ah ok, perfect!

matteoscurati replied:
Uhm, now I remember why I had included this fallback. Without it, I get an eslint error "Type 'string | null' is not assignable to type 'string'". I'll try to investigate a bit more!

matteoscurati replied:
Ok, solved!

Prithpal-Sooriya replied:
Yep nice callout!

danjm commented on app/scripts/controllers/push-platform-notifications/push-platform-notifications.ts:
```diff
@@ -0,0 +1,212 @@
+  },
+};
+
+export class PushPlatformNotificationsController extends BaseController<
```
Could you add a brief code comment that describes the purpose of the controller?

danjm commented on app/scripts/controllers/push-platform-notifications/types/on-chain-notification/schema.d.ts:
```diff
@@ -0,0 +1,300 @@
+/**
+ * This file was auto-generated by openapi-typescript.
```
Perhaps include one more line here on how to generate the file, maybe even just the specific command that was used?

danjm commented on app/scripts/controllers/push-platform-notifications/push-platform-notifications.ts:
```diff
@@ -0,0 +1,212 @@
+  public async enablePushNotifications(UUIDs: string[]) {
+    const bearerToken = await this.getBearerToken();
+    if (!bearerToken) {
+      log.error('Failed to enable push notifications: BearerToken token is missing.');
```
Any reason not to include the `log.error` message in the new Error thrown below? I think that with the current implementation, Sentry won't capture the specific error message.

matteoscurati replied:
No, you're right. I've corrected it using this pattern:

```
      const errorMessage = `Failed to update triggers for push notifications: ${error}`;
      log.error(errorMessage);
      throw new Error(errorMessage);
```

Can it work?

danjm reviewed (commented):
Okay, I have completed a first review of this PR. Looks really good overall. My comments are relatively minor

danjm reviewed (changes requested):
Okay, I have completed a first review of this PR. Looks really good overall. My comments are relatively minor

matteoscurati committed:
docs: :package: add licences and sources for the FCM libraries

matteoscurati committed:
feat: :construction: code improvements after a PR review

matteoscurati commented:
@danjm  thank you very much!

I've reviewed your comments and modified the code as requested. I hope everything is ok.

If so, I'll wait to follow the order for merging the PRs we discussed to:
- change the status of the PR from draft to ready to be merged
- fix the build tests

rekmarks commented on app/scripts/controllers/push-platform-notifications/push-platform-notifications.ts:
```diff
@@ -0,0 +1,229 @@
+import {
```
Why is this controller implemented directly in the extension as opposed to `/core`?

matteoscurati replied:
Yes, we've already discussed this issue for all the other controllers we're using to import the new feature dedicated to notifications. You're absolutely right, and ultimately, we want to implement this and the other controllers in core, but currently, for reasons of speed in implementation, we've decided to work directly in the extension

matteoscurati committed:
feat: :rotating_light: fix lint errors

matteoscurati committed:
feat: :green_heart: set the env variables in the build file

matteoscurati committed:
feat: :recycle: import firebase using npm

matteoscurati committed:
allow-scripts and add Lavamoat policies

matteoscurati committed:
feat: :green_heart: Add Lavamoat policies

matteoscurati committed:
Merge branch 'develop' into feature/install-and-use-fcm-sw

danjm reviewed (dismissed):
All of my comments have been addressed. I am approving. We will need to get reviews from other codeowners.

matteoscurati committed:
Merge branch 'develop' into feature/install-and-use-fcm-sw

matteoscurati committed:
Merge branch 'develop' into feature/install-and-use-fcm-sw

matteoscurati committed:
Merge branch 'develop' into feature/install-and-use-fcm-sw

matteoscurati committed:
Merge branch 'develop' into feature/install-and-use-fcm-sw

matteoscurati committed:
refactor: :sparkles: move the logic to handle push notifications to be ready for MV3

matteoscurati committed:
feat: :sparkles: remove the fcmToken from the state

matteoscurati committed:
Merge branch 'develop' into feature/install-and-use-fcm-sw

matteoscurati committed:
test: :green_heart: fix CI tests

matteoscurati committed:
feat: :green_heart: fix lavamoat policies

matteoscurati committed:
Merge branch 'develop' into feature/install-and-use-fcm-sw

matteoscurati committed:
feat: :green_heart: fix lavamoat policies

Prithpal-Sooriya committed:
build: add global env for push notifications

also update lavamoat policies again

Prithpal-Sooriya committed:
build: update lavamoat again

matteoscurati committed:
Merge branch 'develop' into feature/install-and-use-fcm-sw

matteoscurati committed:
Merge branch 'develop' into feature/install-and-use-fcm-sw

matteoscurati committed:
Merge branch 'develop' into feature/install-and-use-fcm-sw

danjm commented on app/scripts/metamask-controller.js:
```diff
@@ -2849,6 +2864,9 @@ export default class MetamaskController extends EventEmitter {
     delete flatState.vault;
 
+    // The FCM should not be exposed to the UI
+    delete flatState.fcmToken;
```
I think we can avoid having to do this. If the `PushPlatformNotificationsController` is added to `this.store` in the metamask-controller but not `this.memStore`, then it's state won't be exposed to the UI, but it will still be persisted.

matteoscurati replied:
Ok, thanks for the comment, now everything is much clearer. Removed. Thx!

matteoscurati committed:
Merge branch 'develop' into feature/install-and-use-fcm-sw

matteoscurati committed:
refactor: :recycle: keep the fcmtoken in the state

Prithpal-Sooriya reviewed (commented):
Some comments - will resolve them

Prithpal-Sooriya committed:
refactor: suggested changes

remove duplicate files and files that should not impact this build

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/constants/notification-schema.ts:
```diff
@@ -33,6 +33,16 @@ export const NOTIFICATION_CHAINS = {
   LINEA: '59144',
 };
 
+export const CHAIN_SYMBOLS = {
+  [NOTIFICATION_CHAINS.ETHEREUM]: 'ETH',
+  [NOTIFICATION_CHAINS.OPTIMISM]: 'ETH',
+  [NOTIFICATION_CHAINS.BSC]: 'BNB',
+  [NOTIFICATION_CHAINS.POLYGON]: 'MATIC',
+  [NOTIFICATION_CHAINS.ARBITRUM]: 'ETH',
+  [NOTIFICATION_CHAINS.AVALANCHE]: 'AVAX',
+  [NOTIFICATION_CHAINS.LINEA]: 'ETH',
+};
```
Hardcoded Chain Symbols since we need Platform agnostic values (not tied to extension).

If there are shared values then sweet, otherwise we'll keep as is for now.

Prithpal-Sooriya committed:
refactor: fix linting

Prithpal-Sooriya committed:
refactor: revert file changes that are not needed

Prithpal-Sooriya committed:
refactor: remove controller comment

this isn't really needed - we can add a feature comment once everything is merged

Prithpal-Sooriya commented on lavamoat/browserify/beta/policy.json:
```diff
@@ -3748,6 +3748,121 @@
         "setTimeout": true
       }
     },
+    "firebase": {
```
A good chunk of firebase permissions.

It sucks that the npm module is a "large" npm module, when realistically we only use App and Messaging.

These do get tree shaken, but guessing lavamoat policies run on the node modules?

Prithpal-Sooriya reviewed (approved):
LGTM! We will wait for an extension review (and also lets merge this after next RC to avoid additional MV3 debugging)

matteoscurati committed:
Merge branch 'develop' into feature/install-and-use-fcm-sw

danjm reviewed (approved):
LGTM

matteoscurati committed:
Merge branch 'develop' into feature/install-and-use-fcm-sw
